README.md
Signature Pad - Lightweight
===========================

Signature Pad is a JavaScript library for drawing smooth signatures. It's HTML5 canvas based and uses variable width Bézier curve interpolation based on [Smoother Signatures](http://corner.squareup.com/2012/07/smoother-signatures.html) post by [Square](https://squareup.com).
It works in all modern desktop and mobile browsers and doesn't depend on any external libraries.

This is a new fork based on [szimek/signature_pad](https://github.com/szimek/signature_pad) following these goals: 
* minimal footprint - only the `app.js` and `signature_pad.js` files and one CSS formatted element `signature-pad canvas` is needed
* also an absolutely minium example with just the basic elements
* rewritten drawing logic to make it similar to the [Android version](https://github.com/stephanbrunker/android-signaturepad), much more tidier source code and fixed some issues like drawing dots and the draw start sequence. 

![Example](https://user-images.githubusercontent.com/18150071/30086866-aaceae46-929d-11e7-936a-fb830d36bd9f.JPG)

## Demo
[Demo](http://stephanbrunker.github.io/signature_pad) works in desktop and mobile browsers. You can check out its [source code](https://github.com/stephanbrunker/signature_pad/blob/master/js/signature_pad.js). 

For some tips on how to handle window resize and high DPI screens you can also find more about in [HTML5 Rocks tutorial](http://www.html5rocks.com/en/tutorials/canvas/hidpi) or the [base](https://github.com/szimek/signature_pad) of this fork. There you can also get a much more styled version using CSS and more info to use the library on runtime with JavaScript.

## Usage

### CSS,HTML & PHP
Copy the signature_pad.js and app.js files into the /js/ directory on your site

The signature is drawn on a canvas, this is the only element you need to format in width and height which gives you the size of the signature.
 
In the PHP/html form, you include that canvas inside a div with id *"signature-pad"* and an hidden input named *img-data* where the script dumps the image. And of course, you need to link to the script.

```php
<!-- Form for getting signature -->
<form ... >
	<div id='signature-pad' class='signature-pad'>
	  <canvas>This browser doesn't support the canvas for the signaturepad.</canvas>
	  <div class='signature-pad-footer' >
		  <button class="left" id="save" type="button" >Save</button>
		  <button class="right" id="clear" type="button" >Clear</button>
	  </div>
	</div>
<input type="hidden" name="img-data" id="img-data"> 
</form>
<noscript><p>This site needs JavaScript to work properly</p></noscript>
<script src="js/signature_pad.js"></script> 
<script src="js/app.js"></script> 

<!-- show the signature you got as POST -->
<?php 
$img_data = $_POST['img-data']; // contains the image as dataURL
?>
<p><img src='<?php echo $img_data;?>' height='150px' width='400px' alt='signature'></p> 
```

### Options
<dl>
<dt>minWidth</dt>
<dd>(float) Minimum width of a line. Defaults to <code>1</code>.</dd>
<dt>maxWidth</dt>
<dd>(float) Maximum width of a line. Defaults to <code>2.5</code>.</dd>
<dt>backgroundColor</dt>
<dd>(string) Color used to clear the background. Can be any color format accepted by <code>context.fillStyle</code>. Defaults to <code>"rgba(0,0,0,0)"</code> (transparent black). Use a non-transparent color e.g. <code>"rgb(255,255,255)"</code> (opaque white) if you'd like to save signatures as JPEG images.</dd>
<dt>penColor</dt>
<dd>(string) Color used to draw the lines. Can be any color format accepted by <code>context.fillStyle</code>. Defaults to <code>"black"</code>.</dd>
<dt>velocityFilterWeight</dt>
<dd>(float) Weight used to modify new velocity based on the previous velocity. Defaults to <code>0.7</code>.</dd>
</dl>

## License
Released under the [MIT License](http://www.opensource.org/licenses/MIT).


CHANGELOG.md
## Changelog
### 2.0 new Fork by stephanbrunker 
* In this completely revised version of the signature-pad for Javascript I did these changes:
* First, I sorted and commented the different functions for better insight. 
* Then I rewrote most of drawing logic. The original version didn't like dots, so I sorted it out following the cleanest way to do it. In Javascript, the DOWN event has the same coordinates as the first MOVE, and the last MOVE has the same as the UP event, even if there is only a single click, an MOVE event is fired. So, i add points at MOVE events, and at the UP event i decide: If the points array has four members, a bezier curve is drawn until the 3rd point, so I double the last point to draw it to the last position. Three members aren't possible, because then the first point would be doubled to start the bezier. With two members, you can only draw a single line, so I added a function for this and with only one point you need a dot - I added this also. 
* consolidated some of the functions to make it cleaner.
* the Pad works now exactly the same way as my fork of the Android version https://github.com/stephanbrunker/android-signaturepad
* cleaned out the repository with only the essentials
* I made up a new fork because my pull request to the upstream repo was denied - perhaps too many changes ...

### 1.5.3
* Fix `touchend` event on touch devices. (#150) [mtomic](https://github.com/mtomic)
* Fix handling touch events in Egde browser. (#134) [dideldum73](https://github.com/dideldum73)

### 1.5.2
* Prevent loading an empty string in `fromDataURL`. (#108) [Remo](https://github.com/Remo)
* Reject points generated by resting hand (better handling of multi touch). (#48 and #57) [jurreantonisse](https://github.com/jurreantonisse)

### 1.5.1
* Prevent duplicate events on tap in iOS Safari. [PerfectPixel](https://github.com/PerfectPixel)

### 1.5.0
* Add `on` method that rebinds all event handlers. [Alplob](https://github.com/Alplob)

### 1.4.0
* Add `off` method that unbinds all event handlers. [Rob-ot](https://github.com/Rob-ot)

### 1.3.6
* Fix support for Browserify. [chevett](https://github.com/chevett)

### 1.3.5
* Add support for CommonJS/AMD/UMD.

### 1.3.4
* Really fix `fromDataURL` on HiDPI screens.

### 1.3.3
* Fix `fromDataURL` on HiDPI screens.

### 1.3.2
* Fix `onBegin` and `onEnd` callbacks when passed as options to constructor. [yinsee](https://github.com/yinsee)

### 1.3.1
* Fix handling touch events on mobile IE. [tocsoft](https://github.com/tocsoft)

### 1.3.0
* Add `onBegin` and `onEnd` callbacks. [rogerz](https://github.com/rogerz)

### 1.2.4
* Fix bug where stroke becomes very thin. [mvirkkunen](https://github.com/mvirkkunen)

### 1.2.3
* Fix `SignaturePad#fromDataURL` on Firefox. [Fr3nzzy](https://github.com/Fr3nzzy)

### 1.2.2
* Make `SignaturePad#isEmpty` return false after loading an image using `SignaturePad#fromDataURL`. [krisivanov](https://github.com/krisivanov)

### 1.2.1
* Fixed `SignaturePad#clear()`.

### 1.2.0
* Add `backgroundColor` option to set custom color of the background on `SignaturePad#clear()`.
* Rename `color` option to `penColor`.
* Fix passing arguments to canvas element on `SignaturePad#toDataURL()`.


js/signature_pad.js
(function (root, factory) {
  if (typeof define === 'function' && define.amd) {
    // AMD. Register as an anonymous module unless amdModuleId is set
    define([], function () {
      return (root['SignaturePad'] = factory());
    });
  } else if (typeof exports === 'object') {
    // Node. Does not work with strict CommonJS, but
    // only CommonJS-like environments that support module.exports,
    // like Node.
    module.exports = factory();
  } else {
    root['SignaturePad'] = factory();
  }
}(this, function () {

/*!
 * Signature Pad v1.5.3
 * https://github.com/szimek/signature_pad
 *
 * Copyright 2016 Szymon Nowak
 * Released under the MIT license
 *
 * The main idea and some parts of the code (e.g. drawing variable width Bézier curve) are taken from:
 * http://corner.squareup.com/2012/07/smoother-signatures.html
 *
 * Implementation of interpolation using cubic Bézier curves is taken from:
 * http://benknowscode.wordpress.com/2012/09/14/path-interpolation-using-cubic-bezier-and-control-point-estimation-in-javascript
 *
 * Algorithm for approximated length of a Bézier curve is taken from:
 * http://www.lemoda.net/maths/bezier-length/index.html
 *
 * A lot of comments added and the functions ordered for better understandability
 * and some minor improvements by Stephan Brunker
 * 
 */
var SignaturePad = (function (document) {
    "use strict";

    var SignaturePad = function (canvas, options) {
        var self = this,
            opts = options || {};

        this.velocityFilterWeight = opts.velocityFilterWeight || 0.7;
        this.minWidth = opts.minWidth || 1;
        this.maxWidth = opts.maxWidth || 2.5;
        
        this.penColor = opts.penColor || "black";
        this.backgroundColor = opts.backgroundColor || "rgba(0,0,0,0)";
        this.onEnd = opts.onEnd;
        this.onBegin = opts.onBegin;

        this._canvas = canvas;
        this._ctx = canvas.getContext("2d");
        this.clear();

        // we need add these inline so they are available to unbind while still having
        //  access to 'self' we could use _.bind but it's not worth adding a dependency
        this._handleMouseDown = function (event) {
            if (event.which === 1) {
                self._mouseButtonDown = true;
                self._strokeBegin(event);
            }
        };

        this._handleMouseMove = function (event) {
            if (self._mouseButtonDown) {
                self._strokeUpdate(event);
            }
        };

        this._handleMouseUp = function (event) {
            if (event.which === 1 && self._mouseButtonDown) {
                self._mouseButtonDown = false;
                self._strokeEnd(event);
            }
        };

        this._handleTouchStart = function (event) {
            if (event.targetTouches.length == 1) {
                var touch = event.changedTouches[0];
                self._strokeBegin(touch);
             }
        };

        this._handleTouchMove = function (event) {
            // Prevent scrolling.
            event.preventDefault();

            var touch = event.targetTouches[0];
            self._strokeUpdate(touch);
        };

        this._handleTouchEnd = function (event) {
            var wasCanvasTouched = event.target === self._canvas;
            if (wasCanvasTouched) {
                event.preventDefault();
                self._strokeEnd(event);
            }
        };

        this._handleMouseEvents();
        this._handleTouchEvents();
    };

	// Event Listener
    SignaturePad.prototype._handleMouseEvents = function () {
        this._mouseButtonDown = false;

        this._canvas.addEventListener("mousedown", this._handleMouseDown);
        this._canvas.addEventListener("mousemove", this._handleMouseMove);
        document.addEventListener("mouseup", this._handleMouseUp);
    };

    SignaturePad.prototype._handleTouchEvents = function () {
        // Pass touch events to canvas element on mobile IE.
        this._canvas.style.msTouchAction = 'none';
		this._canvas.style.touchAction = 'none';

        this._canvas.addEventListener("touchstart", this._handleTouchStart);
        this._canvas.addEventListener("touchmove", this._handleTouchMove);
        this._canvas.addEventListener("touchend", this._handleTouchEnd);
    };

    SignaturePad.prototype.on = function () {
        this._handleMouseEvents();
        this._handleTouchEvents();
    };

    SignaturePad.prototype.off = function () {
        this._canvas.removeEventListener("mousedown", this._handleMouseDown);
        this._canvas.removeEventListener("mousemove", this._handleMouseMove);
        document.removeEventListener("mouseup", this._handleMouseUp);

        this._canvas.removeEventListener("touchstart", this._handleTouchStart);
        this._canvas.removeEventListener("touchmove", this._handleTouchMove);
        document.removeEventListener("touchend", this._handleTouchEnd);
    };

    SignaturePad.prototype.isEmpty = function () {
        return this._isEmpty;
    };

    
    // kind of an constructor for the member variables
    SignaturePad.prototype._reset = function () {
        this.points = [];
        this._lastVelocity = 0;
        this._lastWidth = (this.minWidth + this.maxWidth) / 2;
        this._isEmpty = true;
        this._ctx.fillStyle = this.penColor;
    };

    SignaturePad.prototype.clear = function () {
        var ctx = this._ctx,
            canvas = this._canvas;

        ctx.fillStyle = this.backgroundColor;
        ctx.clearRect(0, 0, canvas.width, canvas.height);
        ctx.fillRect(0, 0, canvas.width, canvas.height);
        this._reset();
    };

	//========================================================================
	
	// Import and Export Bitmap
    SignaturePad.prototype.toDataURL = function (imageType, quality) {
        var canvas = this._canvas;
        return canvas.toDataURL.apply(canvas, arguments);
    };

    SignaturePad.prototype.fromDataURL = function (dataUrl) {
        var self = this,
            image = new Image(),
            ratio = window.devicePixelRatio || 1,
            width = this._canvas.width / ratio,
            height = this._canvas.height / ratio;

        this._reset();
        image.src = dataUrl;
        image.onload = function () {
            self._ctx.drawImage(image, 0, 0, width, height);
        };
        this._isEmpty = false;
    };

	//========================================================================
	//   User-Defined Types
	//========================================================================

	// UDT Variable POINT
    var Point = function (x, y, time) {
        this.x = x;
        this.y = y;
        this.time = time || new Date().getTime();
    };
    
		// Returns the velocity to this point from start point
        Point.prototype.velocityFrom = function (start) {
			if ( this.distanceTo(start) === 0 ) { return 0; }
			else return (this.time !== start.time) ? this.distanceTo(start) / (this.time - start.time) : 1;
		};

		Point.prototype.distanceTo = function (start) {
			return Math.sqrt(Math.pow(this.x - start.x, 2) + Math.pow(this.y - start.y, 2));
		};

	// UDT Variable BEZIER
    var Bezier = function (startPoint, control1, control2, endPoint) {
        this.startPoint = startPoint;
        this.control1 = control1;
        this.control2 = control2;
        this.endPoint = endPoint;
    };
    
		// Returns approximated length.
		Bezier.prototype.length = function () {
			var steps = 10,
				length = 0,
				i, t, cx, cy, px, py, xdiff, ydiff;

			for (i = 0; i <= steps; i++) {
				t = i / steps;
				cx = this._point(t, this.startPoint.x, this.control1.x, this.control2.x, this.endPoint.x);
				cy = this._point(t, this.startPoint.y, this.control1.y, this.control2.y, this.endPoint.y);
				if (i > 0) {
					xdiff = cx - px;
					ydiff = cy - py;
					length += Math.sqrt(xdiff * xdiff + ydiff * ydiff);
				}
				px = cx;
				py = cy;
			}
			return length;
		};

		Bezier.prototype._point = function (t, start, c1, c2, end) {
			return          start * (1.0 - t) * (1.0 - t)  * (1.0 - t)
				   + 3.0 *  c1    * (1.0 - t) * (1.0 - t)  * t
				   + 3.0 *  c2    * (1.0 - t) * t          * t
				   +        end   * t         * t          * t;
		};

	//========================================================================
	//		Processing Drawing events for Input events
	//========================================================================

	// TouchStart or MouseDown
    SignaturePad.prototype._strokeBegin = function (event) {
        this._reset();
        this._strokeUpdate(event);
        // event of MouseDown is same point as first MouseMove
        if (typeof this.onBegin === 'function') {
            this.onBegin(event);
        }
    };

	// TouchMove or MouseMove
    SignaturePad.prototype._strokeUpdate = function (event) {
        var point = this._createPoint(event);
        this._addPoint(point);
    };


	// TouchEnd or MouseUp - end of Stroke
    SignaturePad.prototype._strokeEnd = function (event) {
		// event of Mouse Up is same point as last MouseMove
		// Points Array can have four Members (Bezier drawn) or two (nothing drawn) or one (single Click)
		var points = this.points;
		
		// shift array left if filled
		switch (points.length) {
			case 4:
				// double the last point to finish the stroke
				this._addPoint(points[4]);
				break;
			case 2:
				// connect with a line
				this._drawLine(points[0],points[1]);
				break;
			case 1:
				// single click - draw a dot
				this._drawDot(points[0]);
				break;
		}

        if (typeof this.onEnd === 'function') {
            this.onEnd(event);
        }
    };
    
    SignaturePad.prototype._strokeDot = function (event) {
		var point = this._createPoint(event);
		this._reset();
		
		if (typeof this.onBegin === 'function') {
            this.onBegin(event);
        }
		
		this._drawDot(point);
		
		if (typeof this.onEnd === 'function') {
            this.onEnd(event);
        }
		
	}
	

	//========================================================================
	//   Drawing a Bezier Line
	//========================================================================

	// add a new Point to the points array, calculate the bezier between P1 and P2 
    SignaturePad.prototype._addPoint = function (point) {
        var points = this.points,
            c2, c3,
            curve, tmp;

        points.push(point);

        if (points.length > 2) {
            // To reduce the initial lag make it work with 3 points
            // by copying the first point to the beginning.
            if (points.length === 3) points.unshift(points[0]);

            tmp = this._calculateCurveControlPoints(points[0], points[1], points[2]);
            c2 = tmp.c2;
            tmp = this._calculateCurveControlPoints(points[1], points[2], points[3]);
            c3 = tmp.c1;
            curve = new Bezier(points[1], c2, c3, points[2]);
            
			var velocity, newWidth;
			velocity = points[2].velocityFrom(points[1]);
			velocity = this.velocityFilterWeight * velocity
				+ (1 - this.velocityFilterWeight) * this._lastVelocity;

			newWidth = this._strokeWidth(velocity);
			this._drawCurve(curve, this._lastWidth, newWidth);

			this._lastVelocity = velocity;
			this._lastWidth = newWidth;

            // Remove the first element from the list,
            // so that we always have no more than 4 points in points array.
            points.shift();
        }
    };

	// Draw Curve to Canvas
    SignaturePad.prototype._drawCurve = function (curve, startWidth, endWidth) {
        var ctx = this._ctx,
            widthDelta = endWidth - startWidth,
            drawSteps, width, i, t, tt, ttt, u, uu, uuu, x, y;

        drawSteps = Math.floor(curve.length());
        ctx.beginPath();
        for (i = 0; i < drawSteps; i++) {
            // Calculate the Bezier (x, y) coordinate for this step.
            t = i / drawSteps;
            tt = t * t;
            ttt = tt * t;
            u = 1 - t;
            uu = u * u;
            uuu = uu * u;

            x = uuu * curve.startPoint.x;
            x += 3 * uu * t * curve.control1.x;
            x += 3 * u * tt * curve.control2.x;
            x += ttt * curve.endPoint.x;

            y = uuu * curve.startPoint.y;
            y += 3 * uu * t * curve.control1.y;
            y += 3 * u * tt * curve.control2.y;
            y += ttt * curve.endPoint.y;

            width = startWidth + ttt * widthDelta;
            this._drawPoint(x, y, width);
        }
        ctx.closePath();
        ctx.fill();
    };

	//========================================================================
	//		Drawing a Single Dot
	//========================================================================

    SignaturePad.prototype._drawDot = function (point) {
        var ctx = this._ctx,
          //  dotSize = typeof(this.dotSize) === 'function' ? this.dotSize() : this.dotSize
          dotSize = this.maxWidth;

        ctx.beginPath();
        this._drawPoint(point.x, point.y, dotSize);
        ctx.closePath();
        ctx.fill();
    };

	//========================================================================
	//		Drawing a Line
	//========================================================================

    SignaturePad.prototype._drawLine = function ( startPoint, endPoint ) {
		var ctx = this._ctx,velocity,newWidth,drawSteps,i,x,y, 
			dx = endPoint.x - startPoint.x, dy = endPoint.y - startPoint.y,
			width,widthDelta;
			
		velocity = endPoint.velocityFrom(startPoint);
		velocity = this.velocityFilterWeight * velocity
			+ (1 - this.velocityFilterWeight) * this._lastVelocity;

		newWidth = this._strokeWidth(velocity);
		widthDelta = newWidth - this._lastWidth;
		
		drawSteps = Math.floor(endPoint.distanceTo(startPoint));
		ctx.beginPath();
        for (i = 0; i < drawSteps; i++) {
			x = startPoint.x + (dx * i / drawSteps);
			y = startPoint.y + (dy * i / drawSteps);
			width = this._lastWidth + (widthDelta * i / drawSteps);
			this._drawPoint(x, y, width);
		}
		ctx.closePath();
        ctx.fill();
        
		this._lastVelocity = velocity;
		this._lastWidth = newWidth;			
		
	};
	
	//========================================================================
	//		All drawings are made by lining up a lot of single points
	//========================================================================	
	
    SignaturePad.prototype._drawPoint = function (x, y, size) {
        var ctx = this._ctx;

        ctx.moveTo(x, y);
        ctx.arc(x, y, size, 0, 2 * Math.PI, false);
        this._isEmpty = false;
    };	
	
	
	//========================================================================
	//		Helper Functions
	//========================================================================

	// Event to POINT converter
    SignaturePad.prototype._createPoint = function (event) {
        var rect = this._canvas.getBoundingClientRect();
        return new Point(
            event.clientX - rect.left,
            event.clientY - rect.top
        );
    };

	// ControlPoint Calculation
    SignaturePad.prototype._calculateCurveControlPoints = function (s1, s2, s3) {
        var dx1 = s1.x - s2.x, dy1 = s1.y - s2.y,
            dx2 = s2.x - s3.x, dy2 = s2.y - s3.y,

            m1 = {x: (s1.x + s2.x) / 2.0, y: (s1.y + s2.y) / 2.0},
            m2 = {x: (s2.x + s3.x) / 2.0, y: (s2.y + s3.y) / 2.0},

            l1 = Math.sqrt(dx1*dx1 + dy1*dy1),
            l2 = Math.sqrt(dx2*dx2 + dy2*dy2),

            dxm = (m1.x - m2.x),
            dym = (m1.y - m2.y),

            k = l2 / (l1 + l2),
            cm = {x: m2.x + dxm*k, y: m2.y + dym*k},

            tx = s2.x - cm.x,
            ty = s2.y - cm.y;

        return {
            c1: new Point(m1.x + tx, m1.y + ty),
            c2: new Point(m2.x + tx, m2.y + ty)
        };
    };

	// stroke width calculation from velocity
    SignaturePad.prototype._strokeWidth = function (velocity) {
        return Math.max(this.maxWidth / (velocity + 1), this.minWidth);
    };


    return SignaturePad;
})(document);

return SignaturePad;

}));
